enhancements/namespace-auto-pruning.md

```diff
@@ -0,0 +1,188 @@
+---
+title: Namespace Auto-pruning Policies
+authors:
+  - "@bcaton85"
+reviewers:
+  - "@dmage"
+  - "@Sunandadadi"
+approvers:
+  - "@dmage"
+  - "@Sunandadadi"
+creation-date: 2023-08-15
+last-updated: 2023-08-15
+status: provisional
+---
+
+# Namespace Auto-pruning Policies
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+1. Should there be a dry run feature to test the policy?
+1. Should this only be implemented in the new UI and not the old UI?
+1. Can a user configure multiple policies per namespace?
+1. Can a user specify tags or repositories by pattern?
+1. Do we want the ability to have tiered priority for auto-prune execution?
+
+## Summary
+
+In order for a container registry to run cost effectively unused resources should be deleted to reclaim storage space. This task should be done automatically in Quay by deleting unused tags through a given policy. The policy used should be configured by the user and a variety policy options should be given. Examples of initial policies would be prune tags older than X number of days or prune tags when more than 100 are present.
+
+## Motivation
+
+Reclaiming space is essential in running a cost efficient registry. Having this done automatically within Quay removes the need for manual intervention or custom software.
+
+### Goals
+
+* User is able to provide a time based policy for auto-pruning tags within a namespace
+* User is able to provide a tag number based policy for auto-pruning tags within a namespace
+* The correct tags are deleted when the policy is applied
+* Pruned tags are tracked via usage logs
+
+### Non-Goals
+
+* Allowing users to specify multiple policies per namespace
+* Applying policies based on repo or tag name pattern matching
+* Allowing policy configurations at the repository level
+* User is able to specify whether tags go into the time machine or are immediately deleted
+
+## Proposal
+
+The details in this proposal is only for namespace auto-pruning policies but designed in a way to support repository auto-pruning policies in the future. This proposal is structured in three parts: how policies are managed, how the policies are structured in the DB, and how the policies will be executed via the auto-prune worker.
+
+### Managing Policies
+
+A user is able to configure auto-pruning within the namespace settings. To start, there will be two auto-pruning options:
```

Comment:
What do you think about defining policies in a more flexible way: 

1) Triggering event (e.g. number of tags exceeds X, or quota is reaching  Y% of its limit or EVERYTIME)
2) Deletion criteria (e.g. tag creation date older than X days, last pull date older than Y days, tag name matches pattern or any combination of them or ALL)

That would create a flexible system to mix and match policies together. Initially we would only start with the following:

1) Supported triggers: number of tags exceeds X, every time
2) Deletion criteria: tags with creation date older than Y days, ALL (sorted by creation or last pull date)

The processing would need to be smart enough to understand that the trigger event defines the limit when to stop deleting, e.g. when number of tags does not exceed X anymore or quota fell below Y%.

The MVP criteria could then be met with: the following selection:

> Prune by number of tags, provided the number days in X:
Trigger event: number of tags exceeds X
Deletion criteria: ALL (sorted by creation date)

> Prune tags older than date, provided number of days in X:
Trigger event: every time (every time the worker wakes up)
Deletion criteria: tag has a creation date older than X days

Initially we could make it so that the UI only offers these two fixed combinations behind a simple dropdown for each policy type. This would leave space for more complex policy creation later.

Reply:
We don't currently have a good way of checking a tag's last pull date, short of going through all of the audit logs. Even then, it might not be the case where the entire audit history is being kept. i.e one might be discarding audit logs older that some time range, and you end up losing that information on latest pull time.

Reply:
I think this creates a more complicated UX and implementation. I think it's better to have unique policy pruning methods where we can enforce what the user can specify rather than account for every combination that a user can provide. Also we're running short on time and accounting for this will put the date at risk.

Reply:
I think we can also achieve flexibility once we support multiple policies. At that point these same operations can be done.

Reply:
Great feedback. The pull date is something we would do in the future, it was just another criteria as an illustrative example. I agree with @bcaton85 about the complexity of my proposal and the extensibility of the current proposal. Let's go with the current model.